src/Guide/Api/Utils.hs

```diff
@@ -89,3 +102,47 @@ instance (HasSwagger api, KnownSymbol name, KnownSymbol desc) =>
              symbolVal (Proxy @desc))
             Nothing
      in toSwagger (Proxy @api) & applyTags [tag]
+
+-- | Servant request details, can be captured by adding `RequestDetails :>` to an API branch.
+data RequestDetails = RequestDetails
+  { rdIp        :: Maybe IP  -- request ip address
+  , rdReferer   :: Maybe Text  -- request referrer
+  , rdUserAgent :: Maybe Text  -- request User-Agent
+  } deriving (Show, Generic)
+
+instance FromHttpApiData IP where
+  parseUrlPiece text
```

Comment:
Just use `readEither` instead, it might give a better error message instead of a generic "not a parseable..."